crates/core/component/stake/src/component/action_handler/undelegate.rs

```diff
@@ -56,12 +63,18 @@ impl ActionHandler for Undelegate {
             })?;
         let expected_unbonded_amount = rate_data.unbonded_amount(u.delegation_amount);
 
-        ensure!(
```

Comment:
What do you think of `check_and_execute(...).map_err(|e| error!(?e))` instead? That way we don't need to add individual tracing statements and we just need to focus on making the error descriptive, which we will have to do as a prelude to using error codes in `CheckTx/DeliverTx` ([#3961](https://github.com/penumbra-zone/penumbra/issues/3961))

Reply:
And we can keep the short `ensure!(invariant, msg)` statements

Reply:
i have some conviction that this is a better pattern, and i'd like to try and convince you of why!

emitting a tracing event at the call-site means that we capture all of the span context that was present, precisely at the moment that an invariant was violated. this means that the tracing subscriber can log an error that looks like:

```
  2024-03-28T13:22:12.703847Z ERROR penumbra_stake::component::action_handler::undelegate: undelegation was prepared for a different epoch, from: Epoch { index: 0, start_height: 0 }, current: Epoch { index: 1, start_height: 2 }                                                                                                                                                                                                             
    at crates/core/component/stake/src/component/action_handler/undelegate.rs:32                                                                                                                                        
    in penumbra_transaction::action::Undelegate with idx: 2                                                                                                                                                             
    in penumbra_mock_consensus::block::execute with height: 3, time: 1711632132                                                                                                                                         
    in app_can_undelegate_from_a_validator::executing block with undelegation transaction
```

this context is lost when we emit tracing error events outside of the function from which the error in question originates. often, the errors propagated back up in the consensus service result in a test failure that looks like this:

```
thread 'app_can_undelegate_from_a_validator' panicked at /home/katie/work/penumbra/crates/core/component/stake/src/component/action_handler/undelegate.rs:32:13:
the service panicked
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: consensus service returned error
test app_can_undelegate_from_a_validator ... FAILED
```

when this happens we have only the error message itself to go off of, and have to rebuild that context ourselves by finding that error and manually tracking down call-sites.

_zooming out:_ in my experience operating other rust services, there is another benefit to emitting tracing events: tracing is a ["scoped, **structured** logging and diagnostics system."](https://docs.rs/tracing/latest/tracing/#), emphasis mine. emitting structured logs allows operators to rely on key-value extraction provided by [various](https://www.splunk.com/en_us/blog/tips-and-tricks/key-value-pair-extraction-definition-examples-and-solutions.html) [popular](https://help.sumologic.com/docs/search/search-query-language/parse-operators/parse-keyvalue-formatted-logs/) [observability](https://docs.datadoghq.com/logs/log_configuration/parsing/?tab=matchers#key-value-or-logfmt) [platforms](https://www.elastic.co/guide/en/logstash/current/plugins-filters-kv.html). that allows operators to query historical logs over time, and to write more fine-grained alerting rules.

_zooming back in:_ the downside of being a little more verbose is real :pensive: but i found myself needing these to debug the `app_can_undelegate_from_a_validator.rs` test; a helpful error message can tell you what went wrong, but a scoped diagnostics system can tell you precisely when and why something went wrong. i could tell that i'd violated an invariant, but it was hard to tell which part of the test had tripped it without adding these.

this, in tandem with test code that uses `tracing::Instrument`, can give us more helpful test failures locally that precisely state what part of the test code cause an error (_see the "executing block with undelegation transaction" span above_) and all of the intermediate steps that led us to an error. 

this is all a more verbose telling of https://github.com/penumbra-zone/penumbra/pull/4017/commits/fd4b8da093aa8742daf959d5c60f0f3e8864a49e, from #4017, but i hope it sells you a bit on this :slightly_smiling_face:

Reply:
to your point though, because brevity is a noble thing, what do you think of 232ee82? i also like the idea of `check_and_execute` being a succinct list of invariants we enforce.

i tried pulling that logic out into two short helpers at the bottom of the file, and now we have a nice `check_and_execute` that looks like

```
async fn check_and_execute<S: StateWrite>(&self, mut state: S) -> Result<()> {
    /* ----- checks ------ */
    // NB: These checks all formerly happened in the `check_historical` method, if profiling
    // shows that they cause a bottleneck we could (CAREFULLY) move some of them back.

    // # Invariant: the undelegation must have been prepared for the current epoch.
    self.check_that_undelegate_was_prepared_for_current_epoch(&state)
        .await?;

    // # Invariant: The unbonded amount must be correctly computed.
    self.check_that_unbonded_amount_is_correctly_computed(&state)
        .await?;

    /* ----- execution ------ */

    // ...

    Ok(())
}
```

...and that feels potentially in the same spirit as the `ensure!(invariant, msg)` we had prior?

Reply:
This makes sense, thanks for the write up. I think we lose something important by formulating invariants negatively, rather than how they would be written in a spec e.g. "the penalties MUST match". But I agree the observability benefits are worth it. Maybe in the future we will get both, until then this looks good to me.

Reply:
I strongly that we should not do this refactor, for a few different reasons.

1. We should avoid doing restructuring of the consensus critical application logic at this stage of development without a strongly motivated reason.

2. We should not move action handling logic out of the action handlers.

3. We should not be coupling our application logic to the way we emit tracing events. This is an internal implementation detail.

Reply:
sounds good. reverted in 7bcd43c56.

Reply:
reverted in b513ddbcfc7bd44117eeffbb4815403b8306d108